// @framer/qoofi2gq.tds-components/@tossteam/colors/adaptive.ts
const adaptive = {
  grey50: 'var(--adaptiveGrey50)',
  grey100: 'var(--adaptiveGrey100)',
  grey200: 'var(--adaptiveGrey200)',
  grey300: 'var(--adaptiveGrey300)',
  grey400: 'var(--adaptiveGrey400)',
  grey500: 'var(--adaptiveGrey500)',
  grey600: 'var(--adaptiveGrey600)',
  grey700: 'var(--adaptiveGrey700)',
  grey800: 'var(--adaptiveGrey800)',
  grey900: 'var(--adaptiveGrey900)',
  opacity50: 'var(--adaptiveOpacity50)',
  opacity100: 'var(--adaptiveOpacity100)',
  opacity200: 'var(--adaptiveOpacity200)',
  opacity300: 'var(--adaptiveOpacity300)',
  opacity400: 'var(--adaptiveOpacity400)',
  opacity500: 'var(--adaptiveOpacity500)',
  opacity600: 'var(--adaptiveOpacity600)',
  opacity700: 'var(--adaptiveOpacity700)',
  opacity800: 'var(--adaptiveOpacity800)',
  opacity900: 'var(--adaptiveOpacity900)',
  background: 'var(--adaptiveBackground)',
  greyBackground: 'var(--adaptiveGreyBackground)',
  layeredBackground: 'var(--adaptiveLayeredBackground)',
  floatBackground: 'var(--adaptiveFloatBackground)',
};

export default adaptive;
