// @framer/qoofi2gq.tds-components/code/Image.tsx
import { addPropertyControls, ControlType } from 'framer';
import * as React from 'react';
import { TDSComponent } from '../src/components/TDSComponent';
import { makeComponentDisplayName } from '../src/utils/makeDisplayName';

interface Props {
  imageURL: string;
  imageFilePath: string;
}

export function Image({ imageURL, imageFilePath }: Props) {
  const src = imageFilePath ? imageFilePath : imageURL;
  return (
    <TDSComponent
      name="Image"
      summary={{
        imageURL: src,
      }}
    >
      <img src={src} style={{ width: '100%', height: 'auto' }} />
    </TDSComponent>
  );
}

addPropertyControls(Image, {
  imageURL: {
    title: '이미지주소',
    type: ControlType.String,
    defaultValue: 'https://static.toss.im/illusts/img_thanksgiving_push_1.png',
  },
  imageFilePath: {
    title: '업로드',
    type: ControlType.File,
    allowedFileTypes: ['jpg', 'jpeg', 'png', 'gif'],
  },
});

(Image as any).defaultProps = {
  width: 200,
  height: 201,
};

Image.displayName = makeComponentDisplayName('Image', '이미지');
