// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Keypad/AlphabetKeypad.tsx
import * as React from 'react';

import VirtualKeypad from './VirtualKeypad';

type AlphabetKey =
  | 'A'
  | 'a'
  | 'B'
  | 'b'
  | 'C'
  | 'c'
  | 'D'
  | 'd'
  | 'E'
  | 'e'
  | 'F'
  | 'f'
  | 'G'
  | 'g'
  | 'H'
  | 'h'
  | 'I'
  | 'i'
  | 'J'
  | 'j'
  | 'K'
  | 'k'
  | 'L'
  | 'l'
  | 'M'
  | 'm'
  | 'N'
  | 'n'
  | 'O'
  | 'o'
  | 'P'
  | 'p'
  | 'Q'
  | 'q'
  | 'R'
  | 'r'
  | 'S'
  | 's'
  | 'T'
  | 't'
  | 'U'
  | 'u'
  | 'V'
  | 'v'
  | 'W'
  | 'w'
  | 'X'
  | 'x'
  | 'Y'
  | 'y'
  | 'Z'
  | 'z';

export interface Props {
  /**
   * 알파벳 키패드 배열
   * - A B C D E F G H I J K L M N O P Q R S T U V W X Y Z
   */
  alphabets: AlphabetKey[];

  /**
   * Key를 클릭하면 발생하는 이벤트
   *
   * @param value string
   */
  onKeyClick: (value: string) => void;

  /**
   * Backspace를 클릭하면 발생하는 에빈트
   */
  onBackspaceClick: () => void;
}

export default class AlphabetKeypad extends React.PureComponent<Props> {
  public static defaultProps = {
    alphabets: [
      'A',
      'B',
      'C',
      'D',
      'E',
      'F',
      'G',
      'H',
      'I',
      'J',
      'K',
      'L',
      'M',
      'N',
      'O',
      'P',
      'Q',
      'R',
      'S',
      'T',
      'U',
      'V',
      'W',
      'X',
      'Y',
      'Z',
    ],
  };

  public render() {
    const { alphabets: characters, onKeyClick, onBackspaceClick } = this.props;
    return (
      <VirtualKeypad
        characters={characters}
        onKeyClick={onKeyClick}
        onBackspaceClick={onBackspaceClick}
        numberOfColumns={7}
      />
    );
  }
}
