// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Keypad/NumberKeypad.tsx
import * as React from 'react';

import VirtualKeypad from './VirtualKeypad';

type NumberKey = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export interface Props {
  /**
   * 숫자 키패드의 순서
   * - 1, 2, 3, 4, 5, 6, 7, 8, 9, 0
   */
  numbers: NumberKey[];

  /**
   * Key를 클릭하면 발생하는 이벤트
   *
   * @param value string
   */
  onKeyClick: (value: number) => void;

  /**
   * Backspace를 클릭하면 발생하는 에빈트
   */
  onBackspaceClick: () => void;
}

export default class NumberKeypad extends React.PureComponent<Props> {
  public static defaultProps = {
    numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 0],
  };

  public onKeyClick = (value: string) => {
    this.props.onKeyClick(parseInt(value, 10));
  };

  public render() {
    const { onBackspaceClick, numbers } = this.props;
    const characters = numbers.map(numberKey => `${numberKey}`);

    return (
      <VirtualKeypad
        numberOfColumns={3}
        characters={characters}
        onKeyClick={this.onKeyClick}
        onBackspaceClick={onBackspaceClick}
      />
    );
  }
}
