// @framer/qoofi2gq.tds-components/src/components/rulers/XRuler.tsx
import styled from '@emotion/styled';
import React from 'react';
import { ZIndex } from '../../constants/zIndex';

export const RULER_COLOR = 'orange';

interface Props {
  color?: string;
  labelOffset?: number;
  distance: number;
  dimension: {
    top: string;
    right: string;
    bottom: string;
    left: string;
  };
}

export function XRuler({
  color = RULER_COLOR,
  labelOffset = 4,
  distance,
  dimension: { top, right, bottom, left },
}: Props) {
  return (
    <Wrapper
      color={color}
      style={{
        width: `${distance}px`,
        top,
        right,
        bottom,
        left,
      }}
    >
      {distance > 0 && (
        <Label color={color} labelOffset={labelOffset}>
          {Math.round(distance)}
        </Label>
      )}
    </Wrapper>
  );
}

const Wrapper = styled.div<{ color: string }>`
  position: absolute;
  height: 2px;
  background: ${({ color }) => color};
  z-index: ${ZIndex.Ruler};
`;

const Label = styled.span<{ color: string; labelOffset: number }>`
  position: relative;
  text-align: center;
  top: -10px;
  left: ${({ labelOffset }) => labelOffset}px;
  font-size: 13px;
  line-height: 20px;
  padding: 2px;
  background: ${({ color }) => color};
  color: white;
  font-weight: 700;
`;
