// @framer/qoofi2gq.tds-components/src/components/rulers/YRuler.tsx
import styled from '@emotion/styled';
import React from 'react';
import { ZIndex } from '../../constants/zIndex';
import { RULER_COLOR } from './XRuler';

interface Props {
  color?: string;
  labelOffset?: number;
  distance: number;
  dimension: {
    top: string;
    right: string;
    bottom: string;
    left: string;
  };
}

export function YRuler({
  color = RULER_COLOR,
  labelOffset = 4,
  distance,
  dimension: { top, right, bottom, left },
}: Props) {
  return (
    <Wrapper
      color={color}
      style={{
        height: `${distance}px`,
        top,
        right,
        bottom,
        left,
      }}
    >
      {distance > 0 && (
        <Label labelOffset={labelOffset} color={color}>
          {Math.round(distance)}
        </Label>
      )}
    </Wrapper>
  );
}

const Wrapper = styled.div<{ color: string }>`
  position: absolute;
  width: 2px;
  background: ${({ color }) => color};
  z-index: ${ZIndex.Ruler};
`;

const Label = styled.span<{ color: string; labelOffset: number }>`
  display: inline-block;
  width: 30px;
  text-align: center;
  position: relative;
  top: ${({ labelOffset }) => labelOffset}px;
  left: -15px;
  font-size: 13px;
  background: ${({ color }) => color};
  color: white;
  font-weight: 700;
`;
