// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Dialog/Dialog.tsx
import cx from "classnames";
import * as React from "react";
import { ComponentProps, ReactNode } from "react";
import Modal from "../Modal";
import DialogButton from "./DialogButton";
import DialogCloseButton from "./DialogCloseButton";
import DialogCTAButton from "./DialogCTAButton";
import DialogIcon from "./DialogIcon";
import DialogImage from "./DialogImage";
import DialogText from "./DialogText";
import DialogTitle from "./DialogTitle";
import DialogWithCTA from "./DialogWithCTA";

type ModalProps = Pick<
  ComponentProps<typeof Modal>,
  "open" | "onEntered" | "onExited" | "closeOnDimmerClick"
>;

interface Props extends ModalProps {
  children: ReactNode;
  onClose?: () => void;
  buttons?: ReactNode;
  className?: string;
  textAlign?: "left" | "center";
}

function Dialog({
  children,
  onClose,
  buttons = (
    <DialogButton onClick={onClose} style="weak" type="primary">
      닫기
    </DialogButton>
  ),
  className,
  textAlign = "left",
  ...rest
}: Props) {
  return (
    <Modal
      className={cx(
        "modal-dialog__card",
        `modal-dialog__card--align-${textAlign}`,
        className
      )}
      {...rest}
    >
      <div className="modal-dialog__body">{children}</div>
      <div className="modal-dialog__button-group">{buttons}</div>
    </Modal>
  );
}

Dialog.Image = DialogImage;
Dialog.Text = DialogText;
Dialog.Title = DialogTitle;
Dialog.Button = DialogButton;
Dialog.Icon = DialogIcon;
Dialog.WithCTA = DialogWithCTA;
Dialog.CTAButton = DialogCTAButton;
Dialog.CloseButton = DialogCloseButton;

export default Dialog;
