// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/ListRow/ListRowTexts.types.tsx
import { ReactElement } from "react";
import { EssentialTextProps } from "./ListRowTexts";

export interface Texts1RowProps {
  top: ReactElement | string;
  topProps?: EssentialTextProps;
}

interface Texts1RowTypeA extends Texts1RowProps {
  type: "1RowTypeA";
}

interface Texts1RowTypeB extends Texts1RowProps {
  type: "1RowTypeB";
}

interface Texts1RowTypeC extends Texts1RowProps {
  type: "1RowTypeC";
}

interface TextsRight1RowTypeA extends Texts1RowProps {
  type: "Right1RowTypeA";
}

interface TextsRight1RowTypeB extends Texts1RowProps {
  type: "Right1RowTypeB";
}

interface TextsRight1RowTypeC extends Texts1RowProps {
  type: "Right1RowTypeC";
}

interface TextsRight1RowTypeD extends Texts1RowProps {
  type: "Right1RowTypeD";
}

export interface Texts2RowProps extends Texts1RowProps {
  bottom: ReactElement | string;
  bottomProps?: EssentialTextProps;
}

interface Texts2RowTypeA extends Texts2RowProps {
  type: "2RowTypeA";
}

interface Texts2RowTypeB extends Texts2RowProps {
  type: "2RowTypeB";
}

interface Texts2RowTypeC extends Texts2RowProps {
  type: "2RowTypeC";
}

interface Texts2RowTypeD extends Texts2RowProps {
  type: "2RowTypeD";
}

interface Texts2RowTypeE extends Texts2RowProps {
  type: "2RowTypeE";
}

interface Texts2RowTypeF extends Texts2RowProps {
  type: "2RowTypeF";
}

interface TextsRight2RowTypeA extends Texts2RowProps {
  type: "Right2RowTypeA";
}

interface TextsRight2RowTypeB extends Texts2RowProps {
  type: "Right2RowTypeB";
}

interface TextsRight2RowTypeC extends Texts2RowProps {
  type: "Right2RowTypeC";
}

interface TextsRight2RowTypeD extends Texts2RowProps {
  type: "Right2RowTypeD";
}

interface TextsRight2RowTypeE extends Texts2RowProps {
  type: "Right2RowTypeE";
}

export interface Texts3RowProps extends Texts2RowProps {
  middle: ReactElement | string;
  middleProps?: EssentialTextProps;
}

interface Texts3RowTypeA extends Texts3RowProps {
  type: "3RowTypeA";
}

interface Texts3RowTypeB extends Texts3RowProps {
  type: "3RowTypeB";
}

export type ListRowTextsProps =
  | Texts1RowTypeA
  | Texts1RowTypeB
  | Texts1RowTypeC
  | TextsRight1RowTypeA
  | TextsRight1RowTypeB
  | TextsRight1RowTypeC
  | TextsRight1RowTypeD
  | Texts2RowTypeA
  | Texts2RowTypeB
  | Texts2RowTypeC
  | Texts2RowTypeD
  | Texts2RowTypeE
  | Texts2RowTypeF
  | TextsRight2RowTypeA
  | TextsRight2RowTypeB
  | TextsRight2RowTypeC
  | TextsRight2RowTypeD
  | TextsRight2RowTypeE
  | Texts3RowTypeA
  | Texts3RowTypeB;
