// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/CardBackgrounds/CardBackground.tsx
import classNames from 'classnames';
import * as React from 'react';

import useResizeObserver from '../../hooks/useResizeObserver';
import generateID from '../../utils/generateID';
import { BezierMask } from '../BezierMask.tsx';

// 테스트 해봤을 때 corner-radius 가 bezier path 를 침범하기 시작하는 값들 보고 정했습니다.
// 예)
// 12px 이면 10px 빼야됨
// 20px 이면 3px 빼야됨
const RADIUS_OFFSET_RATIO = 6;

export const CardBackground: React.FC<{
  className?: string;
  radius?: number;
}> = React.memo(({ className, children, radius = 12 }) => {
  const ref = React.useRef<HTMLDivElement | null>(null);
  const id = React.useRef(generateID('cardbackground'));
  const [rect, setRect] = React.useState<{ width: number; height: number } | undefined>(undefined);

  useResizeObserver(ref, ({ clientHeight, clientWidth }) => {
    if (rect === undefined || rect.height !== clientHeight || rect.width !== clientWidth) {
      setRect({
        width: clientWidth,
        height: clientHeight,
      });
    }
  });

  return (
    <div className="card-background">
      <section
        ref={ref}
        style={{
          clipPath: `url(#${id.current})`,
          borderRadius: `${radius - parseInt(`${radius / RADIUS_OFFSET_RATIO}`)}px`,
        }}
        className={classNames(className, 'card-background--card')}
      >
        {children}
        {rect !== undefined ? (
          <BezierMask id={id.current} width={rect.width} height={rect.height} radius={radius} />
        ) : null}
      </section>
    </div>
  );
});
