// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Loader/Loader.tsx
import classnames from "classnames";
import * as React from "react";

export interface Props {
  className?: string;
  size?: "small" | "medium" | "large";
  type?: "dark" | "light" | "primary";
  label?: string;
  style?: React.CSSProperties;
}

export const Loader: React.FC<Props> = React.memo(
  ({ type = "primary", size = "large", className, label, style }) => {
    return (
      <div
        className={classnames(
          "loader",
          `loader--type-${type}`,
          `loader--size-${size}`,
          className
        )}
        style={style}
      >
        <svg
          className="loader__animation"
          width="48"
          height="48"
          viewBox="0 0 66 66"
        >
          <circle
            className="loader__animation__circle"
            fill="none"
            strokeWidth="6"
            strokeLinecap="round"
            cx="33"
            cy="33"
            r="30"
          />
        </svg>
        {label !== undefined ? (
          <span className="loader__label typography-t6">{label}</span>
        ) : null}
      </div>
    );
  }
);

export default Loader;
