// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Error/Error.tsx
import * as React from 'react';

import convertNewLineToJSX from '../../utils/convertNewLineToJSX';
import { Button } from '../Button';
import { Icon } from '../Icon';

export interface Props {
  /**
   * 아이콘의 이름
   */
  iconName?: string;

  /**
   * 에러 페이지 헤딩
   */
  title?: string;

  /**
   * 에러 페이지 설명
   */
  description?: string;

  /**
   * 버튼에 들어갈 텍스트
   */
  buttonText?: string;

  /**
   * 버튼 클릭 이벤트
   */
  onButtonClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

const Error: React.SFC<Props> = ({ iconName, title, description, buttonText, onButtonClick }) => (
  <div className="error-wrapper">
    <div className="error-wrapper__container">
      {iconName !== undefined ? (
        <Icon className="error-wrapper__icon" name={iconName} size={24} />
      ) : (
        ''
      )}
      {title !== undefined ? (
        <h1 className="error-wrapper__title font-size--17">{convertNewLineToJSX(title)}</h1>
      ) : (
        ''
      )}
      {description !== undefined ? (
        <p className="error-wrapper__description font-size--15">
          {convertNewLineToJSX(description)}
        </p>
      ) : (
        ''
      )}
      {buttonText !== undefined ? (
        <Button type="primary" size="medium" onClick={onButtonClick}>
          {buttonText}
        </Button>
      ) : (
        ''
      )}
    </div>
  </div>
);

export default Error;
