// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/GridList/GridList02.tsx
import cx from "classnames";
import * as React from "react";
import { ReactNode } from "react";
import GridListItem02 from "./GridListItem02";

export interface Props {
  /**
   * 컴포넌트의 root element(`div`)에 추가되는 className입니다.
   */
  className?: string;

  /**
   * 표시될 열의 개수를 정합니다.
   */
  column?: 2 | 3;

  /**
   * @default 16
   */
  verticalPadding?: 0 | 8 | 16;

  /**
   * false로 주면 divider가 사라집니다.
   * @default true
   */
  border?: boolean;

  children?: ReactNode;
}

function GridList02({
  className,
  column = 3,
  verticalPadding = 16,
  border = true,
  children,
}: Props) {
  return (
    <div
      className={cx(
        "grid-list",
        {
          [`grid-list--type-02`]: border,
        },
        `grid-list--column-${column}`,
        `grid-list--column-gap-8`,
        `grid-list--row-gap-${verticalPadding}`,
        className
      )}
    >
      <ul className="grid-list__list">{children}</ul>
    </div>
  );
}

GridList02.Item = GridListItem02;

export default GridList02;
