// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/ListHeader/BaseListHeader.tsx
import classnames from "classnames";
import * as React from "react";
import { LiHTMLAttributes, ReactNode } from "react";

export interface Props extends LiHTMLAttributes<HTMLLIElement> {
  /**
   * 상단 여백의 높이를 지정합니다.
   */
  topPadding?: 0 | 8 | 16 | 24;

  /**
   * 하단 여백의 높이를 지정합니다.
   */
  bottomPadding?: 0 | 8 | 16 | 24;

  /**
   * 하단 구분선을 설정합니다.
   */
  border?: "full" | "indented" | "none";

  right?: ReactNode;
}

function BaseListHeader({
  className,
  topPadding = 24,
  bottomPadding = 8,
  border = "full",
  children,
  right,
  ...rest
}: Props) {
  return (
    <li
      className={classnames(
        "list-header",
        `padding-top-${topPadding}`,
        `padding-bottom-${bottomPadding}`,
        {
          [`has-hairline-after`]: border !== "none",
          [`has-hairline-after--padding24`]: border === "indented",
        },
        className
      )}
      {...rest}
    >
      <div className="list-header__content">{children}</div>
      {right !== undefined ? (
        <div className="list-header__right">{right}</div>
      ) : null}
    </li>
  );
}

export default BaseListHeader;
