// @framer/qoofi2gq.tds-components/src/components/ChildrenPlaceholder.tsx
import styled from '@emotion/styled';
import * as React from 'react';
import { colors } from '../../@tossteam/colors';

interface Props extends React.HTMLAttributes<HTMLDivElement> {
  className?: string;
}

export const ChildrenPlaceholder: React.FC<Props> = ({ className, ...rest }: Props) => {
  return (
    <Container className={className} {...rest}>
      <Description>
        오른쪽의 <ConnectorIcon /> 모양을 잡고,
        <br />
        다른 Frame에 연결해보세요
      </Description>
    </Container>
  );
};

const Container = styled.div`
  height: 100%;
  background-color: ${colors.blue50};
  border-radius: 6px;

  display: flex;
  align-items: center;
  justify-content: center;
`;

const Description = styled.div`
  color: ${colors.blue500};
  font-size: 15px;
  letter-spacing: 0px;
  line-height: 1.5;
  font-weight: 600;
  text-align: center;
`;

const ConnectorIcon = styled.div`
  display: inline-block;
  width: 12px;
  height: 12px;
  background-color: white;
  border: 3px solid #0099fe;
  border-radius: 50%;
`;
