// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/NewTop/Top.tsx
import classnames from "classnames";
import * as React from "react";

import TopText1 from "./TopText1";
import TopText2 from "./TopText2";
import TopText3 from "./TopText3";
import TopText4 from "./TopText4";
import TopText5 from "./TopText5";
import TopText6 from "./TopText6";

type DivProps = React.HTMLAttributes<HTMLDivElement>;

export interface Props extends DivProps {
  lower?: React.ReactNode;
  upper: React.ReactNode;
  horizontalPadding?: 0 | 24;
  topPadding?: 0 | 24;
}

export function Top({
  className,
  lower,
  upper,
  horizontalPadding = 24,
  topPadding = 24,
  ...rest
}: Props) {
  return (
    <div
      className={classnames("m-top", className, {
        [`top--horizontal-padding-${horizontalPadding}`]:
          typeof horizontalPadding === "number",
        [`top--top-padding-${topPadding}`]: typeof topPadding === "number",
      })}
      {...rest}
    >
      {upper}
      {lower}
    </div>
  );
}

Top.Text1 = TopText1;
Top.Text2 = TopText2;
Top.Text3 = TopText3;
Top.Text4 = TopText4;
Top.Text5 = TopText5;
Top.Text6 = TopText6;

export default Top;
