// @framer/qoofi2gq.tds-components/code/ZStack.tsx
import { addPropertyControls, ControlType } from 'framer';
import * as React from 'react';
import { makeComponentDisplayName } from '../src/utils/makeDisplayName';

interface Props {
  content1: React.ReactNode;
  content2: React.ReactNode;
  content3: React.ReactNode;
  content4: React.ReactNode;
  content5: React.ReactNode;
  content6: React.ReactNode;
}

export function ZStack({ content1, content2, content3, content4, content5, content6 }: Props) {
  const contents = [content1, content2, content3, content4, content5, content6];

  return (
    <div style={{ position: 'relative' }}>
      {contents.map(content =>
        content ? <div style={{ position: 'absolute', width: '100%', height: '100%' }}>{content}</div> : null
      )}
    </div>
  );
}

addPropertyControls(ZStack, {
  content1: {
    title: '컨텐츠 1',
    type: ControlType.ComponentInstance,
  },
  content2: {
    title: '컨텐츠 2',
    type: ControlType.ComponentInstance,
  },
  content3: {
    title: '컨텐츠 3',
    type: ControlType.ComponentInstance,
  },
  content4: {
    title: '컨텐츠 4',
    type: ControlType.ComponentInstance,
  },
  content5: {
    title: '컨텐츠 5',
    type: ControlType.ComponentInstance,
  },
});

(ZStack as any).defaultProps = {
  width: 375,
  height: 667,
};

ZStack.displayName = makeComponentDisplayName('ZStack', 'Z스택');
