// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/TabController/TabController.tsx
import * as React from 'react';

import Tab from '../Tab';
import TabItem from './TabItem';
import TabPanel from './TabPanel';

export interface Props {
  selectedIndex: number;
  tabItems?: TabItem[];

  onTabChange: (index: number, lastIndex: number) => void;
}

export default class TabController extends React.Component<Props> {
  public render() {
    const { tabItems = [] } = this.props;

    return (
      <>
        <Tab onChange={this.handleTabChange}>
          {tabItems.map((item, index) => (
            <Tab.Item key={item.title} selected={this.props.selectedIndex === index}>
              {item.title}
            </Tab.Item>
          ))}
        </Tab>
        <section>
          {tabItems.map((item, index) => (
            <TabPanel key={item.key} selected={this.props.selectedIndex === index}>
              {item.content}
            </TabPanel>
          ))}
        </section>
      </>
    );
  }

  private handleTabChange = (tabIndex: number) => {
    this.props.onTabChange(tabIndex, this.props.selectedIndex);
  };
}
