// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/BottomSheet/Select.tsx
import classnames from "classnames";
import * as React from "react";
import generateID from "../../utils/generateID";
import Checkbox from "../Checkbox";
import { List } from "../List";
import ListRow from "../ListRow";

interface Option {
  name: string;
  value: string;
}

export interface Props {
  className?: string;
  options: Option[];
  value: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export default class BottomSheetSelect extends React.Component<Props> {
  public state: {
    uniqId?: string;
  } = {
    uniqId: undefined,
  };

  public componentDidMount() {
    this.setState({ uniqId: generateID("botton-sheet__select-") });
  }

  public render() {
    const { className, value, options, onChange } = this.props;

    return (
      <List
        className={classnames("bottom-sheet__select", className)}
        headerSeparator="none"
        rowSeparator="none"
      >
        {options.map((option) => {
          const checkboxId = `bottom-sheet__select-${this.state.uniqId}-${option.value}`;
          return (
            <label
              className="bottom-sheet__select-label"
              key={option.value}
              htmlFor={checkboxId}
              role="listitem"
            >
              <ListRow
                verticalPadding="small"
                contents={<ListRow.Texts type="1RowTypeA" top={option.name} />}
                right={
                  <Checkbox
                    onChange={onChange}
                    id={checkboxId}
                    name={this.state.uniqId}
                    value={option.value}
                    checked={option.value === value}
                    type="secondary"
                  />
                }
              />
            </label>
          );
        })}
      </List>
    );
  }
}
