// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Agreement/AgreementRightArrow.tsx
import { adaptive } from "@tossteam/colors";
import classNames from "classnames";
import * as React from "react";
import { HTMLProps } from "react";
import { Icon } from "../Icon";

interface Props
  extends Omit<HTMLProps<HTMLDivElement>, "children" | "onClick"> {
  iconName?: string;
  iconColor?: string;
  iconSize?: number;
  className?: string;
  onClick?(event: React.MouseEvent<HTMLElement>): void;
  children?: never;
}

export default function AgreementRightArrow({
  iconName = "arrow-rightwards",
  iconColor = adaptive.grey400,
  iconSize = 24,
  className,
  onClick,
  ...props
}: Props) {
  const clickable = onClick != null;

  return (
    <div
      role={clickable ? "button" : undefined}
      tabIndex={clickable ? 0 : undefined}
      onClick={onClick}
      className={classNames("agreement-right-arrow", className)}
      {...props}
    >
      <Icon
        name={iconName}
        color={iconColor}
        size={iconSize}
        className="agreement-right-arrow__icon"
      />
    </div>
  );
}
