// @framer/qoofi2gq.tds-components/code/Spline.tsx
import { addPropertyControls, ControlType } from 'framer';
import * as React from 'react';
import { makeComponentDisplayName } from '../src/utils/makeDisplayName';

interface Props {
  link: string;
}

export function Spline({ link }: Props) {
  return link ? (
    <iframe src={link} style={{ width: '100%', height: '100%', border: 'none' }}></iframe>
  ) : (
    <div style={emptyStateStyle}>Spline Export URL을 입력해주세요</div>
  );
}

addPropertyControls(Spline, {
  link: {
    title: '링크',
    placeholder: 'spline.design/…',
    type: ControlType.String,
  },
});

const emptyStateStyle: React.CSSProperties = {
  width: '100%',
  height: '100%',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  borderRadius: 6,
  background: 'rgba(136, 85, 255, 0.3)',
  color: 'rgb(136, 85, 255)',
  border: '1px dashed #85F',
  flexDirection: 'column',
};

(Spline as any).defaultProps = {
  height: 200,
  width: 200,
  tint: '#09F',
};

Spline.displayName = makeComponentDisplayName('Spline', '스플라인(3D)');
