// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/index.ts
import 'es7-object-polyfill';
import Dialog from './components/Dialog';

const DialogWithCTA = Dialog.WithCTA; // Alias

/* Components */
export {
  Agreement,
  AgreementList,
  AgreementRightArrow,
  AgreementRow01,
  AgreementRow02,
  AgreementRow03,
  AgreementRow04,
  AgreementSection,
} from './components/Agreement';
export { default as AmountTop } from './components/AmountTop';
export { default as Badge } from './components/Badge';
export { default as Banner } from './components/Banner';
export { default as BoardRow } from './components/BoardRow';
export { BorderType, default as Border, Props as BorderProps } from './components/Border';
export { default as BottomCTA } from './components/BottomCTA';
export { default as BottomInfo } from './components/BottomInfo';
export { default as BottomSheet } from './components/BottomSheet';
export { default as Bubble } from './components/Bubble';
export { Button, ButtonStyle, ButtonType } from './components/Button';
export * from './components/Calendar';
export * from './components/CardBackgrounds';
// export { default as Carousel } from './components/Carousel';
export { default as Checkbox } from './components/Checkbox';
export { ContainerContext, default as Container } from './components/Container';
export { default as Error } from './components/Error';
export { default as FixedBottomCTA } from './components/FixedBottomCTA';
export { default as Form } from './components/Form';
export { default as FullScreenLoader } from './components/FullScreenLoader';
export * from './components/GridList';
export { Icon, SVGIcon } from './components/Icon';
export { default as IconButton } from './components/IconButton';
export * from './components/InfiniteScroll';
export { AlphabetKeypad, NumberKeypad } from './components/Keypad';
export { List } from './components/List';
export * from './components/ListFooter';
export * from './components/ListHeader';
export { default as ListHeader } from './components/ListHeader';
export { default as ListRow, ListRowProps } from './components/ListRow';
export { default as Loader } from './components/Loader';
export * from './components/LogoSlide';
export { default as Modal } from './components/Modal';
export { default as NewAgreement } from './components/NewAgreement';
export { default as NPS } from './components/NPS';
export { default as AlertDialog } from './components/Overlays/AlertDialog';
export { default as ConfirmDialog } from './components/Overlays/ConfirmDialog';
export * from './components/Post';
export { default as ProgressBar } from './components/ProgressBar';
export { default as Radio } from './components/Radio';
export { default as Result } from './components/Result';
export { default as SearchField } from './components/SearchField';
export { default as Select } from './components/Select';
export { default as Slider } from './components/Slider';
export { default as Switch } from './components/Switch';
export { default as Tab } from './components/Tab';
export * from './components/TableRow';
export * from './components/TDSProvider';
export {
  default as Text,
  default as Txt,
  FontWeight,
  Props as TextProps,
  Typography,
} from './components/Text';
export { default as TextArea } from './components/TextArea';
export { default as TextButton } from './components/TextButton';
export { default as TextField } from './components/TextField';
export { default as TextFieldBig } from './components/TextFieldBig';
export { default as TextFieldLine } from './components/TextFieldLine';
export { default as Toast } from './components/Toast';
export {
  Top01,
  Top01Props,
  Top02,
  Top02Props,
  Top03,
  Top03Props,
  Top04,
  Top04Props,
  Top05,
  Top05Props,
  Top06,
  Top06Props,
} from './components/Top';
export { useBottomSheet } from './providers/BottomSheetContext';
export { useDialog } from './providers/DialogContext';
export { useToast } from './providers/ToastContext';
/* ETC. */
export { default as Color } from './styles/Color';
export { androidSafeAreaPatchCssClass } from './utils/hooks/useSafeAreaPatch';
export { default as isSafeAreaExists } from './utils/isSafeAreaExists';
export { Dialog, DialogWithCTA };
