// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Top/Top.tsx
import classnames from "classnames";
import * as React from "react";

export interface Props {
  /**
   * 컴포넌트의 root element(`div`)에 추가되는 className입니다.
   */
  className?: string;
  /**
   * 컴포넌트의 color, fontWeight, textAlign 을 설정할 수 있습니다.
   */
  style?: Pick<React.CSSProperties, "color" | "fontWeight" | "textAlign">;
  /**
   * @description word-break: keep-all을 적용할지 여부를 나타냅니다.
   * @default true
   */
  wordBreak?: boolean;
  horizontalPadding?: 0 | 24;
  topPadding?: 0 | 16 | 24 | 36;
  bottomPadding?: 0 | 6 | 8 | 16 | 24;
}

export class Top extends React.PureComponent<Props> {
  public render() {
    const {
      className,
      style,
      horizontalPadding,
      topPadding,
      bottomPadding,
      wordBreak = true,
      children,
    } = this.props;

    return (
      <div
        className={classnames(
          "top",
          {
            [`top--horizontal-padding-${horizontalPadding}`]:
              typeof horizontalPadding === "number",
            [`top--top-padding-${topPadding}`]: typeof topPadding === "number",
            [`top--bottom-padding-${bottomPadding}`]:
              typeof topPadding === "number",
            "top--word-break": wordBreak,
          },
          className
        )}
        style={style}
      >
        {children}
      </div>
    );
  }
}
