// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/ListFooter/ListFooter.tsx
import classnames from 'classnames';
import * as React from 'react';

import Without from '../../types/Without';

export interface Props {
  /**
   * 컴포넌트의 root element(`li`)에 추가되는 className입니다.
   */
  className?: string;

  /**
   * 상단 구분선을 설정합니다.
   */
  border?: 'full' | 'indented' | 'none';

  /**
   * `li` element의 html 기본 attributes를 지정합니다.
   */
  liAttributes?: Without<React.LiHTMLAttributes<HTMLLIElement>, keyof Props>;
}

export default class ListFooter extends React.PureComponent<Props> {
  public render() {
    const { className, children, liAttributes = {}, border = 'full' } = this.props;

    return (
      <li
        className={classnames(
          'list-footer',
          {
            'list-footer--clickable': liAttributes.onClick,
            [`list-footer--border-${border}`]: border,
            [`has-hairline-before`]: border !== 'none',
            [`has-hairline-before--padding24`]: border === 'indented',
          },
          className
        )}
        {...liAttributes}
      >
        {children}
      </li>
    );
  }
}
