// @framer/qoofi2gq.tds-components/@tossteam/colors/adaptiveDictionary.ts
import colors from "./colors";

export const adaptiveDictionary = {
  grey50: [colors.grey50, colors.inverseGrey50],
  grey100: [colors.grey100, colors.inverseGrey100],
  grey200: [colors.grey200, colors.inverseGrey200],
  grey300: [colors.grey300, colors.inverseGrey300],
  grey400: [colors.grey400, colors.inverseGrey400],
  grey500: [colors.grey500, colors.inverseGrey500],
  grey600: [colors.grey600, colors.inverseGrey600],
  grey700: [colors.grey700, colors.inverseGrey700],
  grey800: [colors.grey800, colors.inverseGrey800],
  grey900: [colors.grey900, colors.inverseGrey900],
  opacity50: [colors.greyOpacity50, colors.whiteOpacity50],
  opacity100: [colors.greyOpacity100, colors.whiteOpacity100],
  opacity200: [colors.greyOpacity200, colors.whiteOpacity200],
  opacity300: [colors.greyOpacity300, colors.whiteOpacity300],
  opacity400: [colors.greyOpacity400, colors.whiteOpacity400],
  opacity500: [colors.greyOpacity500, colors.whiteOpacity500],
  opacity600: [colors.greyOpacity600, colors.whiteOpacity600],
  opacity700: [colors.greyOpacity700, colors.whiteOpacity700],
  opacity800: [colors.greyOpacity800, colors.whiteOpacity800],
  opacity900: [colors.greyOpacity900, colors.whiteOpacity900],
  background: [colors.background, colors.darkBackground],
  greyBackground: [colors.greyBackground, colors.darkGreyBackground],
  layeredBackground: [colors.layeredBackground, colors.darkLayeredBackground],
  floatBackground: [colors.floatBackground, colors.darkFloatBackground],
};

export default adaptiveDictionary;
