// @framer/qoofi2gq.tds-components/src/components/iPhoneSafeAreaContainer.tsx
import styled from '@emotion/styled';
import * as React from 'react';
import { colors } from '../../@tossteam/colors';

interface Props extends React.HTMLAttributes<HTMLDivElement> {
  iPhoneX: boolean;
  safeAreaHeight?: number;
  backgroundColor?: string;
  safeAreaBackgroundColor?: string;
}

export const IPhoneSafeAreaContainer: React.FC<Props> = ({
  children,
  iPhoneX,
  safeAreaHeight = 34,
  backgroundColor = 'transparent',
  safeAreaBackgroundColor = 'transparent',
  ...rest
}) => {
  return (
    <Container backgroundColor={backgroundColor} {...rest}>
      <ChildrenContainer>{children}</ChildrenContainer>
      {iPhoneX && (
        <HomeBarArea height={safeAreaHeight} backgroundColor={safeAreaBackgroundColor}>
          <HomeBar />
        </HomeBarArea>
      )}
    </Container>
  );
};

const Container = styled.div<{ backgroundColor: string }>`
  background-color: ${({ backgroundColor }) => backgroundColor};
  height: 100%;

  display: flex;
  flex-direction: column;
`;

const ChildrenContainer = styled.div`
  position: relative;
  transform: translateZ(0);
  flex: 1;

  display: flex;
  flex-direction: column;
`;

const HomeBarArea = styled.div<{ height: number; backgroundColor: string }>`
  position: relative;
  height: ${({ height }) => height}px;
  background-color: ${({ backgroundColor }) => backgroundColor};
  flex: none;

  display: flex;
  flex-direction: column-reverse;
  align-items: center;
`;

const HomeBar = styled.div`
  margin-bottom: 8px;
  width: 134px;
  height: 5px;
  background-color: ${colors.grey900};
  border-radius: 6px;
`;
