// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Overlays/ConfirmDialog.tsx
import * as React from 'react';
import { Children, cloneElement, ReactElement, ReactNode, useRef, useState } from 'react';
import { animated, useSpring } from 'react-spring';
import { adaptive } from '../../../../../@tossteam/colors';
import { isStaticRenderTarget } from '../../../../../src/utils/isStaticRenderTarget';
import Txt from '../../components/Text';
import convertNewLineToJSX from '../../utils/convertNewLineToJSX';
import ResizedMask from '../ResizedMask/ResizedMask';

interface Props {
  open: boolean;
  title: ReactNode;
  description?: ReactNode;
  onClose?: () => void;
  cancelButton: ReactElement;
  confirmButton: ReactElement;
  parentHeight?: number;
  disableAnimation?: boolean;
}

function ConfirmDialog({
  open,
  title,
  description,
  onClose,
  cancelButton,
  confirmButton,
  parentHeight,
  disableAnimation = false,
}: Props) {
  const bodyRef = useRef<HTMLDivElement>(null);
  const [height, setHeight] = useState<number | undefined>(undefined);
  const [isDisplayed, setIsDisplayed] = useState(false);
  const [isRested, setIsRested] = useState(false);
  const bodyHeight = isStaticRenderTarget() ? bodyRef.current?.clientHeight ?? 0 : height;

  const containerHeight = isStaticRenderTarget() ? parentHeight ?? 0 : window.innerHeight;

  const spring = useSpring({
    opacity: open ? 1 : 0,
    translateY: open ? -(containerHeight / 2 + (bodyHeight ?? 0) / 2) : 0,
    config: {
      tension: 300,
      friction: 26,
      precision: 0.01,
    },
    immediate: disableAnimation,
    onFrame: ({ opacity: progress }: { opacity: number }) => {
      if ((open && progress > 0.95) || (!open && progress < 0.05)) {
        setIsDisplayed(open);
      }
    },
    onRest: () => {
      setIsRested(true);
    },
  });

  const cancelButtonProps = Children.only(cancelButton).props;

  const dialogBody = (
    <animated.div
      className="dialog-wrapper"
      style={{
        transform: spring.translateY?.interpolate(v => `translate(-50%, ${v}px)`),
      }}
      ref={bodyRef}
    >
      <div className="dialog confirm-dialog">
        <Txt typography="t4" fontWeight="bold" color={adaptive.grey800} as="h3">
          {typeof title === 'string' ? convertNewLineToJSX(title) : title}
        </Txt>
        {description !== undefined ? (
          <Txt typography="t6" fontWeight="medium" className="dialog__description" color={adaptive.grey700} as="p">
            {typeof description === 'string' ? convertNewLineToJSX(description) : description}
          </Txt>
        ) : null}
        <div className="confirm-dialog__buttons">
          {cloneElement(cancelButton, {
            type: 'dark',
            style: 'weak',
            ...cancelButtonProps,
          })}
          {confirmButton}
        </div>
      </div>
    </animated.div>
  );

  return (
    <>
      <div className="dialog-area" style={!isDisplayed && !open ? { display: 'none' } : {}}>
        <animated.div
          className="dimmer"
          style={{
            position: isStaticRenderTarget() ? 'absolute' : undefined,
            opacity: spring.opacity,
            pointerEvents: open && isDisplayed ? undefined : 'none',
          }}
          onClick={event => {
            if (event.target !== event.currentTarget) {
              return;
            }

            onClose?.();
          }}
        >
          {isStaticRenderTarget() ? (
            dialogBody
          ) : (
            <ResizedMask
              radius={isRested && open ? 16 : 0}
              onResize={({ height }) => {
                setHeight(height);
              }}
            >
              {dialogBody}
            </ResizedMask>
          )}
        </animated.div>
      </div>
      {!isStaticRenderTarget() && height === undefined ? (
        <div
          ref={elem => {
            if (height !== undefined || elem == null) {
              return;
            }

            setHeight(elem.clientHeight);
          }}
          style={{
            position: 'absolute',
            top: '-9999px',
          }}
        >
          {dialogBody}
        </div>
      ) : null}
    </>
  );
}

export default ConfirmDialog;
