// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/types/OverridableComponenet.ts
import * as React from "react";

/**
 * A component whose root component can be controlled via a `component` prop.
 *
 * Adjusts valid props based on the type of `component`.
 */
export interface OverridableComponent<M extends OverridableTypeMap> {
  <C extends React.ReactType>(
    props: {
      /**
       * The component used for the root node.
       * Either a string to use a HTML element or a component.
       */
      component: C;
    } & OverrideProps<M, C>
  ): JSX.Element;
  (props: DefaultComponentProps<M>): JSX.Element;
}

/**
 * Props of the component if `component={Component}` is used.
 */
export type OverrideProps<
  M extends OverridableTypeMap,
  C extends React.ReactType
> = BaseProps<M> & Omit<React.ComponentPropsWithRef<C>, keyof BaseProps<M>>;

/**
 * Props if `component={Component}` is NOT used.
 */
export type DefaultComponentProps<M extends OverridableTypeMap> = BaseProps<M> &
  Omit<React.ComponentPropsWithRef<M["defaultComponent"]>, keyof BaseProps<M>>;

/**
 * Props defined on the component (+ common material-ui props).
 */
export type BaseProps<M extends OverridableTypeMap> = M["props"] & CommonProps;

/**
 * Props that are valid for material-ui components.
 */
export interface CommonProps {
  className?: string;
  style?: React.CSSProperties;
}

export interface OverridableTypeMap {
  props: {};
  defaultComponent: React.ReactType;
}
