// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/List/List.tsx
import classnames from "classnames";
import * as React from "react";
import { HTMLAttributes } from "react";

export interface Props extends HTMLAttributes<HTMLUListElement> {
  /**
   * 리스트의 `header` 엘리먼트의 기본 하단 구분자를 지정합니다.
   */
  headerSeparator?: "full" | "indented" | "none";

  /**
   * 리스트의 `row` 엘리먼트의 기본 하단 구분자를 지정합니다.
   */
  rowSeparator?: "full" | "indented" | "none";

  /**
   * 리스트 위, 아래 패딩 값을 조절합니다.
   */
  verticalPadding?: string;
}

function List({
  className,
  headerSeparator,
  rowSeparator,
  verticalPadding,
  style,
  ...ulAttributes
}: Props) {
  return (
    <ul
      className={classnames(
        "t-list",
        {
          [`list--header-separator-${headerSeparator}`]: headerSeparator,
          [`list--row-separator-${rowSeparator}`]: rowSeparator,
        },
        className
      )}
      style={{
        paddingTop: verticalPadding,
        paddingBottom: verticalPadding,
        ...style,
      }}
      {...ulAttributes}
    />
  );
}

export default List;
