// @framer/qoofi2gq.tds-components/@tossteam/tds-web/react/components/Badge/Badge.tsx
import classnames from "classnames";
import * as React from "react";
import { ReactNode, useRef, useState } from "react";
import generateID from "../../utils/generateID";
import { BezierMask } from "../BezierMask.tsx";

export interface Props {
  className?: string;
  children?: ReactNode;

  /**
   * `Badge`의 스타일 (모양)
   */
  style?: "fill" | "weak";

  /**
   * `Badge`의 타입 (색상)
   */
  type?: "blue" | "teal" | "green" | "red" | "yellow" | "elephant";

  /**
   * `Badge`의 사이즈
   */
  size?: "large" | "medium" | "small";

  /**
   * 버튼 클릭 이벤트
   */
  onClick?: (event: React.MouseEvent<HTMLSpanElement>) => void;
}

export function Badge({
  style = "fill",
  type = "blue",
  size = "small",
  className,
  onClick,
  children,
}: Props) {
  const id = useRef(generateID("badge"));
  const ref = useRef<HTMLSpanElement | null>(null);
  const [width, setWidth] = useState<number | undefined>(undefined);
  const [height, setHeight] = useState<number | undefined>(undefined);

  // useResizeObserver(ref, ({ clientWidth, clientHeight }) => {
  //   if (width !== clientWidth) {
  //     setWidth(clientWidth);
  //   }

  //   if (height !== clientHeight) {
  //     setHeight(clientHeight);
  //   }
  // });

  return (
    <span
      ref={ref}
      className={classnames(
        `badge`,
        `badge--style-${style}`,
        `badge--type-${type}`,
        `badge--size-${size}`,
        className
      )}
      onClick={onClick}
      style={{
        clipPath: `url(#${id.current})`,
      }}
    >
      {/* Safari에서 clip-path가 content-box처럼 동작하는 이슈를 해결하기 위해서 따로 content를 둔다.
      @see https://stackoverflow.com/questions/54186766/css-clip-path-on-mobile-safari-acts-like-content-box */}
      <span className="badge__content" style={{ whiteSpace: "nowrap" }}>
        {children}
      </span>
      {width !== undefined && height !== undefined ? (
        <BezierMask id={id.current} width={width} height={height} radius={12} />
      ) : null}
    </span>
  );
}

export default Badge;
